sf-stream/src/runtime.rs

```diff
@@ -0,0 +1,227 @@
+            .context
+            .get_transactions(self.current_version, batch_size, ledger_info.version())
+        {
+            Ok(transactions) => {
+                if transactions.is_empty() {
+                    debug!("[sf-stream] no transactions to send");
+                    sleep(Duration::from_millis(100)).await;
+                    return result;
+                }
+                // TODO: there might be an off by one (tx version fetched)
+                debug!(
+                    "[sf-stream] got {} transactions from {} to {} [{}]",
+                    transactions.len(),
+                    self.current_version,
+                    self.current_version + transactions.len() as u64,
+                    transactions.last().map(|txn| txn.version).unwrap_or(0)
+                );
+                for onchain_txn in transactions {
+                    // TODO: assert txn.version == &self.current_version + 1?
+                    // TODO: return a `Result` & check to ensure we pushed before incrementing
+                    let txn_version = onchain_txn.version;
+                    // Todo: since the timestamp is per block, only calculate this value once per block
+                    let timestamp = self
+                        .context
+                        .get_block_timestamp(onchain_txn.version)
+                        .unwrap_or_else(|_| {
+                            panic!(
+                                "Could not get timestamp for version {}",
+                                onchain_txn.version
+                            )
+                        });
+                    let txn = self
+                        .resolver
+                        .as_converter(self.context.db.clone())
+                        .try_into_onchain_transaction(timestamp, onchain_txn)
+                        .unwrap_or_else(|e| {
+                            panic!(
+                                "Could not convert onchain transaction version {} into transaction: {:?}",
+                                txn_version, e
+                            )
+                        });
+
+                    let txn_proto = self.convert_transaction(txn);
+                    self.print_transaction(&txn_proto).await;
+                    result.push(txn_proto);
+                    self.current_version = txn_version;
+                }
+                self.current_version += 1;
+            }
+            Err(err) => {
+                error!("[sf-stream] failed to get transactions: {}", err);
+                sleep(Duration::from_millis(100)).await;
+                return result;
+            }
+        }
+        result
+    }
+
+    pub fn maybe_update_from_block_metadata(&mut self, transaction: &Transaction) {
+        if let Transaction::BlockMetadataTransaction(bmt) = transaction {
+            // TODO: ADD BLOCK HEIGHT UPDATES ONCE ITS IN BMT
+            self.block_height += 1;
+            self.current_epoch = bmt.epoch.0
```

Comment:
Block height will never be in the block metadata transaction.  It's in the resource or the events.

Reply:
Didn't you say that it was going to be there at one point? I guess we'll figure it out once it's somewhere. For now this should be good enough?